Add demo (Puzzletron vs Minitron guide) in examples/pruning/ with README and notebooks (#1320)
NVIDIA/TensorRT-Model-Optimizer: examples/pruning/demo/README.md

# How to Reduce Your LLM Size and Efficiency with NVIDIA Model-Optimizer: A Pruning & Distillation Guide

## Table of Contents

> 1.[Introduction](#1-introduction)

Reviewer: Formatting issue because of missing space 
```suggestion
1. [Introduction](#1-introduction)
```

[Part I — Setup & Experiments](#part-i--setup--experiments)

2. [Prerequisites](#2-prerequisites)
3. [Scenario 1: Quick & Reliable Compression](#3-scenario-1-quick--reliable-compression)
4. [Scenario 2: Hardware-Constrained Compression](#4-scenario-2-hardware-constrained-compression)

[Part II — Results, Analysis & Insights](#part-ii--results-analysis--insights)

5. [Head-to-Head: When Does Each Method Win?](#5-head-to-head-when-does-each-method-win)
6. [Distillation: An Impactful Step](#6-distillation-an-impactful-step)
7. [Inference Performance](#7-inference-performance)
8. [Limitations & Practical Tips](#8-limitations--practical-tips)
9. [Open Questions](#9-open-questions)

> 10.[References](#10-references)

Reviewer: ```suggestion
10. [References](#10-references)
```

Throughout this guide, we use [**Qwen3-8B**](https://huggingface.co/Qwen/Qwen3-8B) as our base model — a dense, Transformer-based, decoder-only LLM with 8B parameters and 36 layers. All compressed variants are evaluated with [**MMLU**](https://arxiv.org/abs/2009.03300) **(5-shot)**. Companion Jupyter notebooks are provided so you can reproduce every result on this model end-to-end.

> **MMLU (Massive Multitask Language Understanding)** is a benchmark covering 57 subjects across STEM, humanities, social sciences, and more. Each question is a 4-choice multiple choice problem, giving a random baseline of 25%. The 5-shot variant provides 5 in-context examples before each question.

| | Scenario 1 | Scenario 2 |
|---|---|---|
| **Goal** | Make my general-purpose model smaller and faster, quickly and reliably | Fit a strict hardware memory budget |
| **Usecases** | Create a smaller version of the same model architecture to form a consistent family | Create a single, highly optimized deployment model that fits specific hardware budget |

> | **Compression** | Light/Moderate (10–15% parameter reduction) | Aggressive (>30% memory reduction) |

Reviewer: We can compress by minitron to 25% without any major drop in accuracy

Author: Replaced with 10-20%. Above that, on Qwen3-8B, we see a real advantage of using Puzzletron compared to Minitron.

| **Method** | Homogeneous Pruning ([Minitron](https://arxiv.org/abs/2408.11796)) | Heterogeneous NAS-based Pruning ([Puzzletron](https://arxiv.org/html/2411.19146v3)) |
| **Complexity** | Low — Importance-based ranking, uniform pruning | High — Fine-grained NAS search + MIP optimization |
| **Output** | Homogeneous (all Transformer blocks have the same structure) | Heterogeneous architecture (variable layer widths) |

> Both paths are followed by **knowledge distillation**, which recovers accuracy lost during pruning. In our Qwen3-8B experiments, we show that significant recovery is possible with as few as 100 training iterations on a small dataset, though actual recovery will vary by model and compression level.

Reviewer: ```suggestion
Both paths are followed by **knowledge distillation**, which recovers accuracy lost during pruning. In our Qwen3-8B experiments, we show that significant recovery (in MMLU) is possible with as few as 100 training iterations on a small dataset, though actual recovery will vary by model and compression level.
```

    classDef distill fill:#F8CECC,stroke:#B85450,color:#000

    A["Baseline Model<br/>(Qwen3-8B)"]:::model --> EVAL0["Evaluation<br/>(baseline)"]:::eval
    A --> B["Distillation Data Prep<br/>WikiText-103"]:::dataprep
    A --> C{"Compression<br/>Method"}:::compress
    C --> D["Minitron<br/>Homogeneous pruning<br/>(Calibration Data Prep under the hood)"]:::compress
    C --> CAL["Calibration Data Prep<br/>Nemotron-Post-Training-v2"]:::dataprep --> E["Puzzletron<br/>Heterogeneous NAS"]:::compress
    D --> F["Evaluation<br/>(pruned)"]:::eval
    E --> F
    F --> G["Knowledge Distillation<br/>Qwen3-8B -> student"]:::distill
    B --> G
    G --> H["Evaluation<br/>(distilled)"]:::eval
```

### What this guide covers

- **Pruning**: structurally removing neurons, attention heads, or entire layers to produce a smaller model.
- **Distillation**: transferring knowledge from the original (teacher) model to the pruned (student) model to recover accuracy lost during pruning.

### What this guide does NOT cover

- **Quantization**: reducing numerical precision (e.g. FP16 → INT8).
- **Sparsity**: zeroing out weights while keeping the architecture.

> - **MoE and hybrid architectures**: this guide focuses on dense Transformer models. Compression of Mixture-of-Experts (MoE) and hybrid architectures will be covered in a future guide.

Reviewer: We cover pruning + distillation + PTQ in `examples/pruning/minitron` for Nano-v2 (Hybrid) in #1325 and soon will be added for Nano v3 as well. Can you cross link that page once the #1325 PR is merged?

### How to read this guide

- **"I need a smaller and faster general-purpose model — quickly and reliably"** → go to [Scenario 1: Quick & Reliable Compression (Section 3)](#3-scenario-1-quick--reliable-compression).
- **"I must fit a strict memory budget"** → go to [Scenario 2: Hardware-Constrained Compression (Section 4)](#4-scenario-2-hardware-constrained-compression).
- **"Which method should I use?"** → read end-to-end. Section 5 compares both methods head-to-head.

**Guide organization:** This guide is split into two parts. **Part I — Setup & Experiments** (Sections 2–4) covers the technicalities needed to reproduce the experiments: environment setup and step-by-step walkthroughs for each scenario. **Part II — Results, Analysis & Insights** (Sections 5–9) focuses on what those experiments reveal: a head-to-head comparison, a deep dive into distillation, inference benchmarks, practical tips and limitations, and open questions for future work.

> > **Note on reproducibility:** The pruning frameworks and scoring mechanisms used by Minitron and Puzzletron are under active development. As ModelOpt evolves, exact evaluation numbers may differ from one release to another. The trends and comparative insights presented in this guide (which method wins in which regime, how distillation behaves, and the accuracy-compression trade-offs) are expected to remain consistent.

Reviewer: can you mention nemo container + modelopt version used for these experiments. For example `nemo:26.02 + mount nvidia-modelopt 0.43.0`?

---

<a id="part-i--setup--experiments"></a>

## Part I — Setup & Experiments

---

## 2. Prerequisites

### Hardware

All experiments in this guide were run on **2x H200 GPUs**. This can be adapted to different GPU counts and types depending on your setup. Adjust tensor/pipeline parallelism and batch size accordingly.

> **Architecture requirement:** The NeMo container and ModelOpt scripts used in this guide require an **x86-64 (AMD64)** host.

### Container

> **Setup source of truth:** ModelOpt evolves quickly. The instructions below reflect the setup used at the time this guide was written and are provided as a working example. For the most up-to-date container version and installation steps, refer to:

> > [puzzletron/mbridge_distillation/README.md](https://github.com/NVIDIA/Model-Optimizer/blob/feature/puzzletron/examples/puzzletron/mbridge_distillation/README.md)

Reviewer: can you replace `feature/puzzletron` links with `main` as the former is no longer valid.

Also note that we no longer have `examples/puzzletron/mbridge_distillation` folder and instead move it to `examples/megatron_bridge/distill.py` for unified minitron/puzzletron distillation script

> We use the [NVIDIA NeMo Framework Docker container (26.02)](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/nemo?version=26.02.00), which includes all the libraries needed pre-installed (including [Megatron-Bridge](https://github.com/NVIDIA/Model-Optimizer/tree/main/examples/megatron_bridge/README.md) — NVIDIA's library that bridges HuggingFace models with the Megatron-LM training framework, enabling efficient multi-GPU distillation).

Reviewer: ```suggestion
We use the [NVIDIA NeMo Framework Docker container (26.02)](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/nemo?version=26.02.00), which includes all the libraries needed pre-installed (including [Megatron-Bridge](https://github.com/NVIDIA/Model-Optimizer/tree/main/examples/megatron_bridge/README.md) — NVIDIA's library that bridges HuggingFace models with the Megatron-core framework, enabling efficient multi-GPU distillation).
```

You need [Docker](https://docs.docker.com/get-docker/) and the [NVIDIA Container Toolkit](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html) to enable GPU access inside containers.

```bash
docker run \
    --gpus all \
    --shm-size=16GB \
    --net=host \
    --ulimit memlock=-1 \
    --rm -it \
    -w /workspace \
    nvcr.io/nvidia/nemo:26.02 bash
```

### Clone ModelOpt

Clone the ModelOpt repository to access examples and scripts:

```bash
git clone https://github.com/NVIDIA/Model-Optimizer.git /workspace/Model-Optimizer
cd /workspace/Model-Optimizer

> ```

Reviewer: can you clone first and then mount so users can always retain changes in modelopt. See this doc how we recommend docker run: https://github.com/NVIDIA/Model-Optimizer/tree/achidiac/pruning_demo/examples/megatron_bridge#pre-requisites

### Evaluation setup

> Inside the container, install `lm-eval` for benchmark evaluation:

Reviewer: do we need to uninstall `nvidia_lm_eval` as well?

Author: not necessary I beleive, but will uninstall as well just in case

```bash

> rm -rf /opt/venv/lib/python3.12/site-packages/lm_eval

Reviewer: ```suggestion
python -m pip uninstall -y lm_eval
```

> pip install lm-eval==0.4.8

Reviewer 2: Can we create a Dockerfile with all the dependencies and ModelOpt installation and have one-line instruction on running the container?

Author: I did consider this, but I kept it explicit since lm-eval==0.4.8 is basically the only extra dependency here, and a separate Dockerfile would add one more thing to maintain. I also noticed a lot of troubleshooting caused by lm_eval version mismatches (only 0.4.8 works), so I think it can be useful to have it explicit and visible

- **FFN (Feed-Forward Network):** A per-token MLP applied after attention. The intermediate size (12288) controls its capacity. Puzzletron can reduce this per layer; Minitron reduces it uniformly across all layers.

Authenticate with HuggingFace and download Qwen3-8B:

```bash
hf auth login --token <your_token>
hf download Qwen/Qwen3-8B --local-dir /workspace/models/Qwen3-8B
```

### Companion notebooks

| Notebook | Description | Runtime (2x H200) |
|---|---|---|
| [`00_prerequisites.ipynb`](00_prerequisites.ipynb) | Prepare the data + Baseline evaluation | ~15 min |
| [`scenario1_minitron.ipynb`](scenario1_minitron.ipynb) | Scenario 1 — Minitron | ~1h45 |
| [`scenario1_puzzletron.ipynb`](scenario1_puzzletron.ipynb) | Scenario 1 — Puzzletron | ~6h (on first Puzzletron run) |
| [`scenario2_minitron.ipynb`](scenario2_minitron.ipynb) | Scenario 2 — Minitron | ~45 min |
| [`scenario2_puzzletron.ipynb`](scenario2_puzzletron.ipynb) | Scenario 2 — Puzzletron | ~6h15 (on first Puzzletron run) |

From within the container, copy the notebooks to your local workspace:

```bash

> cp -r /workspace/Model-Optimizer/examples/pruning/demo/* /workspace

Reviewer: we should just use docker-mounted path directly so changes are retained outside the docker. If needed we can add `chmod -R 777 Model-Optimizer/examples/pruning/` before docker run

```

If this doesn't work, upload the notebooks manually in Jupyter after launching it.

Finally, start Jupyter Lab:

```bash
cd /workspace
pip install --upgrade ipywidgets notebook
jupyter lab --ip 0.0.0.0 --port=8888 --allow-root
```

---

## 3. Scenario 1: Quick & Reliable Compression

> *"I need a smaller and faster general-purpose model — quickly and reliably."*

> You have a working LLM and want to reduce its size by 10–15% (in number of parameters) for general-purpose tasks. You need a straightforward pipeline with predictable results and a standard, homogeneous model as output. You don't want to invest time experimenting and familiarizing with a complex pipeline or dealing with heterogeneous model formats.

Reviewer: ```suggestion
You have a working LLM and want to reduce its size by upto 25% (in number of parameters) for general-purpose tasks. You need a straightforward pipeline with predictable results and a standard, homogeneous model as output. You don't want to invest time experimenting and familiarizing with a complex pipeline or dealing with heterogeneous model formats.
```

Author: up to 20% for consistency with advanced experiments results.

### 5.2 Why each method wins in its regime

In theory, Minitron is a subset of Puzzletron: any architecture Minitron can find, Puzzletron could also find if its search space were large enough. But the search space must be finite, and expanding it comes with significant compute and complexity costs. This is why both methods have their sweet spot.

**Scenario 1 (moderate compression):**

On MMLU, Minitron outperforms Puzzletron at this level (+3.43pp post-distill). Its uniform pruning directly targets a parameter count, produces a clean architecture in a single step, and avoids the complexity of a full NAS pipeline. At the same time, this advantage is benchmark-dependent: on other benchmarks, Puzzletron could retain more of the teacher's accuracy than Minitron (see [Section 5.4](#54-benchmark-specific-behavior)). This means there is no single-bullet approach: different compression methods have their winning territories even at moderate compression. That said, when factoring in pipeline complexity and the standard model format Minitron produces, it remains the practical default for most general-purpose compression needs at this level.

> Moreover, Minitron can be applied **iteratively**: for example, prune 10%, distill, then prune another 10% and distill again. This staged schedule typically preserves more quality than a single, more aggressive pruning step at the same overall parameter reduction.

Reviewer: ```suggestion
Moreover, Minitron can be applied **iteratively**: for example, prune 20%, distill, then prune another 20% and distill again. This staged schedule typically preserves more quality than a single, more aggressive pruning step at the same overall parameter reduction.
```

**At 80% memory, Minitron wins post-distillation, despite losing pre-distillation.** Before distillation, Puzzletron (0.5910) leads Minitron (0.5084) by +8.3pp. After distillation, Minitron (0.7302) overtakes Puzzletron (0.6921) by +3.8pp. This is a concrete example of the architecture ranking flip described in [Section 6.4](#64-architecture-ranking-can-flip-after-distillation).

**The crossover point lies somewhere between 20% and 38% compression.** Below ~20% compression, Minitron consistently wins post-distillation. Beyond ~38%, Puzzletron pulls decisively ahead. The exact crossover will depend on the model, the distillation budget, and the Puzzletron search space — but this range provides a practical guideline.

### 5.6 Decision rules

| If... | Then use... |
|---|---|
| Compression is <15% and general-purpose | **Minitron** |
| You need a standard/homogeneous model | **Minitron** |
| Compression is >20% | **Puzzletron** |
| You have a hard memory budget | **Puzzletron** |
| You want minimal pipeline complexity | **Minitron** |

> | You need to preserve specific capabilities (e.g. reasoning) | **Puzzletron** |

Reviewer: I dont think this is true. We are seeing reasoning capabilities preserved very well for minitron as well. I think this might be the case pre-distillation though

Author: Yes this is clearer pre-distillation, will remove it.

---

## 8. Limitations & Practical Tips

### 8.1 Limitations of this guide

- **Single base model:** All experiments use Qwen3-8B. Results (especially distillation convergence speed and the crossover point between Minitron and Puzzletron) may differ on other models and model families.
- **Limited benchmarks:** The notebooks reproduce MMLU end-to-end. Supplementary evaluations on HellaSwag and GSM8K (see [Section 5.4](#54-benchmark-specific-behavior)) confirm that the best method is benchmark-dependent, but three benchmarks on one model are not enough to build general per-task guidelines.

> - **Minimal distillation:** 100 iterations on WikiText-103 is a lower bound. Production deployments should use more iterations, larger datasets, and domain-specific data.

Reviewer: You can link the distillation data blend perparation guide I added in #1325

> **Going further:** To explore these questions, see [Advanced Compression Experiments: Results & Insights](advanced_compression_experiments.md), which gathers the results and insights from more sophisticated experiments.

---

## 10. References

- **Minitron:** Sreenivas et al., [*Compact Language Models via Pruning and Knowledge Distillation*](https://arxiv.org/abs/2407.14679), 2024.
- **More Minitron Results:** Sreenivas et al., [*LLM Pruning and Distillation in Practice: The Minitron Approach*](https://arxiv.org/pdf/2408.11796), 2024.
- **Puzzletron:** Bercovich et al., [*Puzzle: Distillation-Based NAS for Inference-Optimized LLMs*](https://arxiv.org/abs/2411.19146), 2024.
- **NVIDIA ModelOpt:** [GitHub Repository](https://github.com/NVIDIA/Model-Optimizer)
- **Llama Puzzletron Tutorial:** [Puzzletron Example on ModelOpt](https://github.com/NVIDIA/Model-Optimizer/blob/feature/puzzletron/examples/puzzletron/README.md)

> - **Model Compression with Megatron-Bridge:** [Megatron-Bridge Examples](https://github.com/NVIDIA/Model-Optimizer/tree/main/examples/megatron_bridge/README.md)

Reviewer: ```suggestion
- **Model Compression and distillation with Megatron-Bridge:** [Megatron-Bridge Examples](https://github.com/NVIDIA/Model-Optimizer/tree/main/examples/megatron_bridge/README.md)
```
